csaf_2.0/prose/csaf-v2-editor-draft.md

```diff
@@ -3352,6 +3350,125 @@ This does not exclude that such a document is also available in an access protec
+
+### 5.1.7 Requirement 7: One folder per year
+
+The CSAF documents must be located within folders named `<YYYY>` where `<YYYY>` is the year given in the value of `/document/tracking/initial_release_date`.
+
+Examples:
+
+```
+2021
+2020
+```
+
+### 5.1.8 Requirement 8: index.txt
+
+The index.txt file within MUST provide a list of all filenames of CSAF documents which are located in the sub-directories with their filenames.
+
+Examples:
+
+```
+2020/example_company_-_2020-yh4711.json
+2019/example_company_-_2019-yh3234.json
+2018/example_company_-_2018-yh2312.json
+```
+
+> This can be used to download all CSAF documents.
+
+### 5.1.9 Requirement 9: changes.csv
+
+The file changes.csv must contain the filename as well as the value of `/document/tracking/current_release_date` for each CSAF document in the sub-directories without a heading; lines must be sorted by the latter one.
+
+Examples:
+
+```
+2020/example_company_-_2020-yh4711.json, "2020-07-01T10:09:07Z"
+2018/example_company_-_2018-yh2312.json, "2020-07-01T10:09:01Z"
+2019/example_company_-_2019-yh3234.json, "2019-04-17T15:08:41Z"
+2018/example_company_-_2018-yh2312.json, "2019-03-01T06:01:00Z"
+```
+
+### 5.1.10 Requirement 10: Directory listings
+
+Directory listing SHALL be enabled to support manual navigation.
+
+### 5.1.11 Requirement 11: ROLIE service document
+
+The ROLIE service document MUST be a JSON file that conforms with [RFC8322].
+
+**TODO: Provide Example**
+
+### 5.1.12 Requirement 12: ROLIE category document
+
+Each ROLIE category document MUST be a JSON file that conforms with [RFC8322].
+
+**TODO: Provide Example**
+
+### 5.1.13 Requirement 13: TLP:AMBER and TLP:RED
+
+CSAF documents labeled TLP:AMBER or TLP:RED MUST be access protected. If they are provided via a webserver this SHALL be done under a different path than for TLP:WHITE, TLP:GREEN and unlabeled CSAF documents. TLS client authentication, access tokens or any other automatable authentication method SHALL be used.
+
+### 5.1.14 Requirement 14: Redirects
+
+Redirects SHOULD NOT be used. If they are inevitable only HTTP Header redirects are allowed.
+
+> Reasoning: Clients, as e.g. `curl`, do not follow any other kind of redirects.
+
+### 5.1.15 Requirement 15: Integrity
+
+All CSAF documents SHALL have at least one hash file computed with a secure cryptographic hash algorithm (e.g. SHA-512 or SHA-3) to ensure their integrity. The filename is constructed by appending the file extension which is given by the algorithm.
+
+MD5 and SHA1 SHOULD NOT be used.
+
+Example:
+
+```
+File name of CSAF document: example_company_-_2019-yh3234.json
+File name of SHA-256 hash file: example_company_-_2019-yh3234.json.sha256
+File name of SHA-512 hash file: example_company_-_2019-yh3234.json.sha512
+```
+
+The file content SHALL start with the first byte of the hexadecimal hash value. Any subsequent data (like a filename) which is optional SHALL be separated by at least one space.
```

Comment:
Instead of doubling the amount of files, could the hash be specified as a second field in the index.txt file with a hash prefix specifying the algorithm used? Something like:

```
2020/example_company_-_2020-yh4711.json sha256:321691c4527cb4510e1365d2a982cd141ddd73c2be7a04893f08e56006a40b2e
2019/example_company_-_2019-yh3234.json sha512:7fe6b968ec273dec239437f43ad7662f24484b1fd401c38509993a2e1e2972d47bb5543f7ba6d7bd41503731641e9200a8dcc49b9875ff53dd64098e902bafb9
2018/example_company_-_2018-yh2312.json sha384:f7f4840727a9f3be895dbe246275506fb550ec9b55f8264db606f2f81b6fd5931f7bb82c338df163730de7de6072f214
```

Reply:
I'm not sure whether that is still as easy to generate/parse as it was intended. However, you do not need to fulfill requirement 8 (index.txt) if you provide a ROLIE document... IMHO having that as a separate file helps to use existing tools.

Reply:
Generation is easy as `sha256sum *.whatever | sed s/^/sha256:/g > index.txt` and the collation of file to hash mappings into one file per "intent" is folklore I think. But, traditionally the hash is noted first and last the resource name (as per the example call I note on a Linux system ...). As an implementer I could easily live with bundled info in files with well-known names. Instead of "per file" one would analyze "per line".

I see the complexity rising though as @tschmidtb51 states. E.g. when sibling files are processed the target is fully specified. In such a bundled file, the interpretation of the target names may deviate from what one finds at the assumed path.

Reply:
> I'm not sure whether that is still as easy to generate/parse as it was intended. However, you do not need to fulfill requirement 8 (index.txt) if you provide a ROLIE document... IMHO having that as a separate file helps to use existing tools.

I think the complexity isn't that much different as Stefan illustrated, it's just appending to a file instead of writing out a new one while iterating over a list of CSAF files.

What existing tools process hashes in separate files?

> I see the complexity rising though as @tschmidtb51 states. E.g. when sibling files are processed the target is fully specified. In such a bundled file, the interpretation of the target names may deviate from what one finds at the assumed path.

I'm not sure I understand this. What do you mean by a sibling file?

Reply:
> What existing tools process hashes in separate files?

I was thinking of `sha256sum` or `sha512sum`... They would use e.g. `sha256sum $i > ${i}.sha256` to generate and `sha256sum -c ${i}.sha256` to check the file. (It is assumed that `$i` is the filename of the current CSAF document.)

IMHO, adding the hashes into the index.txt makes updates a little bit more complex (now you would have to edit a line). I fear that this hinders SME to provide integrity as it is then not only uploading a bunch of files (CSAF+hash file) to the web server.

There are 3 other points I would like to bring up:
1. How would we deal then with providing multiple hashes for one file? 
    Tools might make all kind of assumptions (e.g. sha256 is always in the third field) or need to be specific for each vendor (which we tried to prevent). With separate files, we don't have this issue as everyone can choose the hash according to one's wishes / preferences. When you use a certain file extension, you either get a hash file or not. But you don't have to parse through a long list.
2. Does it scale for CSAF aggregators?
    If a CSAF aggregator has to do this, he basically needs to join all this from different files. I think this is more complex (also with updates and maybe deletions) than "just" copying more files. 
3. Big open source projects are already doing their releases that way, e.g.:
   * https://nextcloud.com/changelog/#latest21
   * https://httpd.apache.org/download.cgi
   * https://www.openssl.org/source/
   * http://www.releases.ubuntu.com/21.04/

Reply:
> * How would we deal then with providing multiple hashes for one file?

True. Though, the ease of use of any hashing algorithm on any modern system makes me think that providing more than one is sort of pointless :-)

> * Does it scale for CSAF aggregators?
>   If a CSAF aggregator has to do this, he basically needs to join all this from different files. I think this is more complex (also with updates and maybe deletions) than "just" copying more files.
> 

I guess so, but I would assume that there will be a lot more CSAF publishers than aggregators so the spec should slightly favor the former when it comes to ease of use.

> * Big open source projects are already doing their releases that way, e.g.:
>   
>   * https://nextcloud.com/changelog/#latest21
>   * https://httpd.apache.org/download.cgi
>   * https://www.openssl.org/source/

All of these distribute just a handful of artifacts for which it makes no sense to produce an index file, so creating separate hash files is really the only option. For projects with 1000s of files though, I would guess single with all hashes may be more common (though I can't find a good example of this).

>   * http://www.releases.ubuntu.com/21.04/

This one actually uses the index with hashes (file `SHA256SUMS`) approach :-)

All in all, I see the use cases for single file hashes but I would still slightly prefer a single index of them. I'm fine keeping this as is though. It would be nice to hear from others on what they find more useful.